File: multivendor-ecommerce/src/main/java/com/ecommerce/platform/service/CartService.java
package com.ecommerce.platform.service;

import com.ecommerce.platform.dto.CartItemResponse;
import com.ecommerce.platform.dto.CartResponse;
import com.ecommerce.platform.entity.Cart;
import com.ecommerce.platform.entity.CartItem;
import com.ecommerce.platform.entity.Product;
import com.ecommerce.platform.entity.User;
import com.ecommerce.platform.exception.ResourceNotFoundException;
import com.ecommerce.platform.repository.CartItemRepository;
import com.ecommerce.platform.repository.CartRepository;
import com.ecommerce.platform.repository.ProductRepository;
import com.ecommerce.platform.security.SecurityUtils;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;

@Service
@RequiredArgsConstructor
public class CartService {

    private final CartRepository cartRepository;
    private final CartItemRepository cartItemRepository;
    private final ProductRepository productRepository;
    private final SecurityUtils securityUtils;

    @Transactional(readOnly = true)
    public CartResponse getMyCart() {
        return toResponse(getCurrentCart());
    }

    @Transactional
    public CartResponse addItem(Long productId, int quantity) {
        Cart cart = getCurrentCart();
        Product product = productRepository.findById(productId)
                .orElseThrow(() -> new ResourceNotFoundException("Product not found with id: " + productId));

        if (!product.isActive()) {
            throw new IllegalStateException("This product is no longer available");
        }

        CartItem existing = cartItemRepository.findByCartIdAndProductId(cart.getId(), productId).orElse(null);
        if (existing != null) {
            existing.setQuantity(existing.getQuantity() + quantity);
            cartItemRepository.save(existing);
        } else {
            CartItem item = CartItem.builder()
                    .cart(cart)
                    .product(product)
                    .quantity(quantity)
                    .build();
            cartItemRepository.save(item);
        }

        return toResponse(getCurrentCart());
    }

    @Transactional
    public CartResponse updateItem(Long productId, int quantity) {
        Cart cart = getCurrentCart();
        CartItem item = cartItemRepository.findByCartIdAndProductId(cart.getId(), productId)
                .orElseThrow(() -> new ResourceNotFoundException("This product is not in your cart"));
        item.setQuantity(quantity);
        cartItemRepository.save(item);
        return toResponse(getCurrentCart());
    }

    @Transactional
    public CartResponse removeItem(Long productId) {
        Cart cart = getCurrentCart();

        CartItem item = cartItemRepository
                .findByCartIdAndProductId(cart.getId(), productId)
                .orElseThrow(() ->
                        new ResourceNotFoundException("This product is not in your cart"));

        cart.getItems().remove(item);
        cartItemRepository.delete(item);

        return toResponse(cart);
    }

    @Transactional
    public void clearCart() {
        Cart cart = getCurrentCart();
        cart.getItems().clear();
        cartRepository.save(cart);
    }

    private Cart getCurrentCart() {
        User user = securityUtils.getCurrentUser();
        return cartRepository.findByUserId(user.getId())
                .orElseThrow(() -> new ResourceNotFoundException("Cart not found for this user"));
    }

    private CartResponse toResponse(Cart cart) {
        List<CartItemResponse> items = cart.getItems().stream()
                .map(item -> CartItemResponse.builder()
                        .productId(item.getProduct().getId())
                        .productName(item.getProduct().getName())
                        .unitPrice(item.getProduct().getPrice())
                        .quantity(item.getQuantity())
                        .subtotal(item.getProduct().getPrice().multiply(BigDecimal.valueOf(item.getQuantity())))
                        .build())
                .toList();

        BigDecimal total = items.stream().map(CartItemResponse::getSubtotal)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        return CartResponse.builder()
                .cartId(cart.getId())
                .items(items)
                .totalAmount(total)
                .build();
    }
}

File: multivendor_app/lib/services/cart_service.dart
import 'package:dio/dio.dart';
import '../core/api_client.dart';
import '../models/cart.dart';

class CartService {
  final Dio _dio = ApiClient.instance.dio;

  Future<CartModel> getMyCart() async {
    try {
      final response = await _dio.get('/cart');
      return CartModel.fromJson(response.data);
    } catch (e) {
      throw ApiClient.instance.mapError(e);
    }
  }

  Future<CartModel> addItem({required int productId, required int quantity}) async {
    try {
      final response =
          await _dio.post('/cart/items', data: {'productId': productId, 'quantity': quantity});
      return CartModel.fromJson(response.data);
    } catch (e) {
      throw ApiClient.instance.mapError(e);
    }
  }

  Future<CartModel> updateItem({required int productId, required int quantity}) async {
    try {
      final response = await _dio.put('/cart/items/$productId', data: {'quantity': quantity});
      return CartModel.fromJson(response.data);
    } catch (e) {
      throw ApiClient.instance.mapError(e);
    }
  }

  Future<CartModel> removeItem(int productId) async {
    try {
      final response = await _dio.delete('/cart/items/$productId');
      return CartModel.fromJson(response.data);
    } catch (e) {
      throw ApiClient.instance.mapError(e);
    }
  }

  Future<void> clearCart() async {
    try {
      await _dio.delete('/cart');
    } catch (e) {
      throw ApiClient.instance.mapError(e);
    }
  }
}
